Haraka 2.8.18 crashes when trying to deliver mail to a particular Go Daddy server

### system info
Haraka | Haraka.js — Version: 2.8.18
 --- | :--- 
Node | v6.12.3
OS | Linux 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
openssl | OpenSSL 1.0.2k-fips  26 Jan 2017

### Expected behavior

### Observed behavior
Haraka crashes after trying to deliver email to a particular Go Daddy SMTP server:
p3plibsmtp03-11.prod.phx3.secureserver.net

### Steps to reproduce
Try to deliver email to p3plibsmtp03-11.prod.phx3.secureserver.net


Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] Looking up A records for: smtp.secureserver.net
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] Attempting to deliver to: 68.178.213.203:25 (0) (2)
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound] [outbound::25:68.178.213.203:undefined:50] dispense() clients=1 available=0
Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] [outbound] [outbound::25:68.178.213.203:undefined:50] createResource() - creating obj - count=1 min=0 max=10
Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] [outbound] created uuid=B2C7C758-6E83-4052-876B-CB2DF5D6F208 host=68.178.213.203 port=25 pool_timeout=50
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound] acquired socket B2C7C758-6E83-4052-876B-CB2DF5D6F208 for outbound::25:68.178.213.203:undefined:50
Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] S: 421 p3plibsmtp03-11.prod.phx3.secureserver.net bizsmtp ESMTP server error (try again later)\r\n
Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] C: RSET
Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] Temp fail for: Upstream error: 421 p3plibsmtp03-11.prod.phx3.secureserver.net bizsmtp ESMTP server error (try again later)
Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] running deferred hooks
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] Temp failing 1521758825136_1521758825136_0_16681_OjI6On_260_outbound.example.com for 64 seconds: Upstream error: 421 p3plibsmtp03-11.prod.phx3.secureserver.net bizsmtp ESMTP server error (try again later)
Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] S: 503 5.5.0 not available\r\n
Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] C: RSET
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] bouncing mail: 503 5.5.0 not available
Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] running bounce hooks
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound] Sending email via params
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound] Created transaction: 476415D0-0ED5-4E2B-BDF0-1755913AA948
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound] Adding missing Message-Id header
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound] Adding missing Date header
Mar 22 17:47:05 outbound haraka[16681]: [ERROR] [-] [core] adding missing results store
Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] running pre_send_trans_email hooks
Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound] Processing domain: foo.com
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] Error: Callback was already called.
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:955:32
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at HMailItem.hmail.next_cb (/usr/lib/node_modules/Haraka/outbound/queue.js:40:9)
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at HMailItem.discard (/usr/lib/node_modules/Haraka/outbound/hmail.js:1241:18)
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at outbound.send_email.origin (/usr/lib/node_modules/Haraka/outbound/hmail.js:1203:22)
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at async.forEachSeries (/usr/lib/node_modules/Haraka/outbound/index.js:283:17)
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:473:16
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at replenish (/usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:993:25)
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at iterateeCallback (/usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:983:17)
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:958:16
Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/outbound/index.js:309:17
Mar 22 17:47:05 outbound haraka[16681]: [NOTICE] [-] [core] Shutting down

The first part is just lame - the server sending to (not sure which
software that is) is broken. They don't support RSET after 4xx. I guess
we'll need to implement something to cope with that. I'll put in another
ticket.

The second is a straight up bug that needs more investigation.

On Fri, Mar 23, 2018 at 8:09 AM, lnedry <notifications@github.com> wrote:

> system info
> Haraka Haraka.js — Version: 2.8.18
> Node v6.12.3
> OS Linux 3.10.0-693.21.1.el7.x86_64 #1
> <https://github.com/haraka/Haraka/pull/1> SMP Wed Mar 7 19:03:37 UTC 2018
> x86_64 x86_64 x86_64 GNU/Linux
> openssl OpenSSL 1.0.2k-fips 26 Jan 2017 Expected behavior Observed
> behavior
>
> Haraka crashes after trying to deliver email to a particular Go Daddy SMTP
> server:
> p3plibsmtp03-11.prod.phx3.secureserver.net
> Steps to reproduce
>
> Try to deliver email to p3plibsmtp03-11.prod.phx3.secureserver.net
>
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1]
> [outbound] Looking up A records for: smtp.secureserver.net
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1]
> [outbound] Attempting to deliver to: 68.178.213.203:25 (0) (2)
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound]
> [outbound::25:68.178.213.203:undefined:50] dispense() clients=1
> available=0
> Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] [outbound]
> [outbound::25:68.178.213.203:undefined:50] createResource() - creating
> obj - count=1 min=0 max=10
> Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] [outbound]
> created uuid=B2C7C758-6E83-4052-876B-CB2DF5D6F208 host=68.178.213.203
> port=25 pool_timeout=50
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound]
> acquired socket B2C7C758-6E83-4052-876B-CB2DF5D6F208 for
> outbound::25:68.178.213.203:undefined:50
> Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL]
> [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] S: 421
> p3plibsmtp03-11.prod.phx3.secureserver.net bizsmtp ESMTP server error
> (try again later)\r\n
> Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL]
> [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] C: RSET
> Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] Temp fail for:
> Upstream error: 421 p3plibsmtp03-11.prod.phx3.secureserver.net bizsmtp
> ESMTP server error (try again later)
> Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1]
> [outbound] running deferred hooks
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1]
> [outbound] Temp failing 1521758825136_1521758825136_0_
> 16681_OjI6On_260_outbound.example.com for 64 seconds: Upstream error: 421
> p3plibsmtp03-11.prod.phx3.secureserver.net bizsmtp ESMTP server error
> (try again later)
> Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL]
> [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] S: 503 5.5.0 not
> available\r\n
> Mar 22 17:47:05 outbound haraka[16681]: [PROTOCOL]
> [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1] [outbound] C: RSET
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1]
> [outbound] bouncing mail: 503 5.5.0 not available
> Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [DD3572D9-D0A6-48F5-8370-EC5208FF9523.1.1]
> [outbound] running bounce hooks
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound]
> Sending email via params
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound]
> Created transaction: 476415D0-0ED5-4E2B-BDF0-1755913AA948
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound]
> Adding missing Message-Id header
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound]
> Adding missing Date header
> Mar 22 17:47:05 outbound haraka[16681]: [ERROR] [-] [core] adding missing
> results store
> Mar 22 17:47:05 outbound haraka[16681]: [DEBUG] [-] [core] running
> pre_send_trans_email hooks
> Mar 22 17:47:05 outbound haraka[16681]: [INFO] [-] [core] [outbound]
> Processing domain: printedcom.com
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] Error: Callback
> was already called.
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> /usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:955:32
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> HMailItem.hmail.next_cb (/usr/lib/node_modules/Haraka/
> outbound/queue.js:40:9)
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> HMailItem.discard (/usr/lib/node_modules/Haraka/outbound/hmail.js:1241:18)
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> outbound.send_email.origin (/usr/lib/node_modules/Haraka/
> outbound/hmail.js:1203:22)
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> async.forEachSeries (/usr/lib/node_modules/Haraka/
> outbound/index.js:283:17)
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> /usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:473:16
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at replenish
> (/usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:993:25)
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> iterateeCallback (/usr/lib/node_modules/Haraka/
> node_modules/async/dist/async.js:983:17)
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> /usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:958:16
> Mar 22 17:47:05 outbound haraka[16681]: [CRIT] [-] [core] at
> /usr/lib/node_modules/Haraka/outbound/index.js:309:17
> Mar 22 17:47:05 outbound haraka[16681]: [NOTICE] [-] [core] Shutting down
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haraka/Haraka/issues/2375>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAobYxNq9DAlgx8cMZ8JzF4ak85NhExVks5thOX9gaJpZM4S4nRT>
> .
>

--- a/outbound/hmail.js
+++ b/outbound/hmail.js
@@ -745,6 +745,10 @@ class HMailItem extends events.EventEmitter {
                     // EHLO command was rejected; fall-back to HELO
                     return send_command('HELO', mx.bind_helo);
                 }
+                if (command === 'rset') {
+                    // Broken server doesn't accept RSET, terminate the connection
+                    return client_pool.release_client(socket, port, host, mx.bind, true);
+                }
                 reason = `${code} ${(extc) ? `${extc} ` : ''}${response.join(' ')}`;
                 if (/^rcpt/.test(command) || command === 'dot_lmtp') {
                     if (command === 'dot_lmtp') last_recip = ok_recips.shift();

